src/checks.py

```diff
@@ -171,14 +224,43 @@ def check_git_revision(state: State) -> Optional[str]:
                 f"--git-dir {state.git_dir}/.git "
                 f"checkout --quiet {state.git_hash}"
             ),
-            (
-                f"diff --recursive {state.git_dir} {state.source_dir} "
-                "--exclude DEPENDENCIES --exclude NOTICE --exclude .git "
-                "--exclude .gitignore --exclude Jenkinsfile --exclude .mvn "
-                "--exclude mvnw --exclude mvnw.cmd"
-            ),
         ]
     )
+    if sh_result is not None:
+        return sh_result
+
+    logging.info("NOTE: The following diff output is only informational.")
+    logging.info("NOTE: The actual verification is done in Python.")
+    sh(f"diff --recursive {state.git_dir} {state.source_dir}")
+
+    diff = filecmp.dircmp(state.git_dir, state.source_dir, ignore=[])
+    errors = []
+
+    # First, check that any files appearing in only one tree are allowed
+    errors += _check_dircmp_only_either_allowed(diff)
+
+    # Then make sure that all files that exist in both places have no diff
+    errors += _check_dircmp_no_diff_files(diff)
+    # And finally that there we could compare all the files
+    errors += _check_dircmp_no_funny_files(diff)
+
+    if errors:
+        errors.append("See above for a full output of diff.")
+        return "\n\n".join(errors)
+    return None
+
+
+def check_blacklisted_files(state: State) -> Optional[str]:
+    blacklist = [
```

Comment:
wonder if we shouldn't include the contents of .gitignore also? dunno if the syntax is crap or not so not a blocking comment

Reply:
The syntax is not super bad, but then there's also the matching logic (leading slash vs no leading slash, globbing). Not to mention, we'd also need to handle non-top-level `.gitignore`s. I'll break this out into a separate issue; might be worth doing, but probably not before the stable Docker env and wrapper script are set up.

Reply:
Hmm, on second thought, we'd probably best do this by calling `git` in some moderately smart way.